File: 03-poo/08-abstraccion/java/soluciones/RetoFiguras.java
public class RetoFiguras {
    static abstract class Figura { abstract double area(); }
    static class Circulo extends Figura { final double radio; Circulo(double radio){this.radio=radio;} double area(){return Math.PI*radio*radio;} }
    static class Rectangulo extends Figura { final double base, altura; Rectangulo(double b,double a){base=b;altura=a;} double area(){return base*altura;} }
    public static void main(String[] args){ Figura[] figuras={new Circulo(2),new Rectangulo(3,4)}; for(Figura f:figuras) System.out.println(f.area()); }
}

File: 03-poo/08-abstraccion/scala/soluciones/RetoFiguras.scala
abstract class FiguraReto:
  def area: Double
class CirculoReto(val radio: Double) extends FiguraReto:
  def area: Double = math.Pi * radio * radio
class RectanguloReto(val base: Double,val altura: Double) extends FiguraReto:
  def area: Double = base * altura

@main def retoFiguras(): Unit = List[FiguraReto](new CirculoReto(2),new RectanguloReto(3,4)).foreach(f => println(f.area))
